### ISSUE TITLE
Slasher: Could not cache justified state balances: could not get ancestor state: unknown block

### ISSUE THREAD

**Issue by prestonvanloon** [prysmaticlabs/prysm#9100]
I needed to upgrade our test server's disk drive as it was beyond 90% full.  Prior to starting the process I deleted some old files, namely the old prysm v1.X.X files up to but not including v1.3.10 and some extraneous tar.gz files, to open up some space.  In addition, I toggled the --historical-slasher-node from TRUE to FALSE.  I did this as a test to learn what I might encounter on the production server when the slasher moves to the beacon node in the next release.  I did not think that I'd draw a fatal or unrecoverable error .  Here is what I found:

```
prysm.sh[23271]: time="2021-06-24 16:42:37" level=info msg="Connected to eth1 proof-of-work chain" endpoint="http://127.0.0.1:8545" prefix=powchain
prysm.sh[23271]: time="2021-06-24 16:42:37" level=info msg="Blockchain data already exists in DB, initializing..." prefix=blockchain
prysm.sh[23271]: time="2021-06-24 16:42:38" level=info msg="Chain genesis time reached" genesisStateRoot=2bb257ca66d05a047a65fe43a5f457b674de445d917cca029efb09b3ba4758c4 genesisTime="2020-11>
prysm.sh[23271]: time="2021-06-24 16:42:38" level=fatal msg="Could not cache justified state balances: could not get ancestor state: unknown block" prefix=blockchain
systemd[1]: beacon-chain.service: Main process exited, code=exited, status=1/FAILURE
systemd[1]: beacon-chain.service: Failed with result 'exit-code'.
```
After several attempts to resolve the error including, upgrading the disk drive size, upgrading to Prysm v1.3.11 from v1.3.10 and adjusting the rpc settings to default in the both the beacon-chain and validator config.yaml files the problem persists.   Below are the configs for the beacon-chain and validator.

1. beacon-chain, validator and eth1(geth) exist on the same server
2. beacon-chain config - historical-slasher-node: true disable-broadcast-slashings: true pyrmont: true http-web3provider: http://localhost:8545
3. validator config - wallet-dir:  "server/directory" pyrmont: true graffiti: "graffiti"
4. Server is an ARM64 (4core/16gb) running ubunut 20.04, geth v1.10.4, prysm v1.3.11

Any help in resolving this issue would be greatly appreciated.

_Originally posted by @cin3rama in https://github.com/prysmaticlabs/prysm/issues/7237#issuecomment-868099021_

**Comment by prestonvanloon** [comment]
This is an issue with the current slasher binary. I suspect that this issue will not manifest with the new integrated `--slasher` in the beacon node. 

@rauljordan does this issue look familiar?

**Comment by prestonvanloon** [comment]
@cin3rama what version of prysm were you running at the time of this error?

**Comment by cin3rama** [comment]
@prestonvanloon It started with 1.3.10 and continued through 1.3.11 (auto upgrade to .11 occurred when I rebooted as part of the disk drive volume increase, noted in my original post). 

You are correct, not very familiar with this version of the slasher.  Also observed and extreme disk write condition, I've attached a graph.  Please note - the drop in writes on the graph is because I shut down the validator and beacon chain.
![reads](https://user-images.githubusercontent.com/3494288/123444847-336f4500-d58c-11eb-9663-cc396f674506.png)

**Comment by cin3rama** [comment]
Following up with some additional information and context.
1. The slasher was running in the original configuration, having its own directory and systemd files.  Slasher systemd config - monitoring-host 0.0.0.0, enable-historical-detection: true, pyrmont: true, accept-terms: true
2. All prysm components, beacon-chain, validator and slasher were shut down since my original issue post.  However, I fired up the beacon-chain today and did not experience the the disk extreme write condition, however, the same fatal error still existed.
3. I toggled the historical-slasher-node and disable-broadcast-slashings (in the beacon-chain.yaml) from true to FALSE in both cases today.  I did experience the disk extreme write condition at that point. 
4. Given the results I shut all prysm components down once again.

I know its not much but hopefully the additional info is of some help.  Thank you for staying on this issue.

**Comment by pschlup** [comment]
I had the exact same error message ("Could not cache justified state balances: could not get ancestor state: unknown block"), however I'm not running a slasher. I suspect your db file was somehow corrupted due to the disk being full.

I solved the issue by moving the db file to another location as a backup (mine was at /var/lib/prysm/.eth2/beaconchaindata/beaconchain.db) and letting the beacon-chain sync again from scratch. After ~23 hours the sync is over and the validators are working again.

It would be great to be able to "rollback" the db by deleting the last few hours or days of data or tell Prysm to sync again from a "last known good" checkpoint. Or maybe it could detect that at startup time automatically and sync from there instead of panicking?

I'm not sure what happened in my case, as the disk is only <30% full. A bug might have caused something bad to be committed to the db, or I might have a bad disk. Mine started failing overnight without any restarts or any other reason I could find in any logs. I'm running arch linux on a dedicated dell i5, plenty of ram and disk and UPS power backup.

I will start backing up the db every week so it doesn't take as long if it happens again or if I need to replace the disk.

**Comment by cin3rama** [comment]
Thanks for the info @pschlup.  I too have resolved the issue by re-syncing the beacon-chain from scratch.  In my case the error presented itself after I toggled a slasher flag in the beacon-chain config as mentioned above.  I also have plenty of computing power and disk space (<50% used) for a full prysm/geth node.

In addition to the re-sync I also removed all slasher flags from the beacon-chain config file, stopped the slasher and removed it from systemd as well as deleted the slasher directory, thus the slasher db as well.  I ran this config for 5 days on my testnet (goerli) node and everything worked as expected including getting a proposal the first day.  

In case anyone is interested I made the above changes running the following: Prsym v1.4.1(beacon-chain & validator), geth 1.10.5 on Ubuntu 20.04.  In addition, I just finished testing the system with geth 1.10.6 without issue.  I have had Prysm v1.4.2 running for about a day so far, again no known issues.
